presto-main/src/main/java/com/facebook/presto/failureDetector/HeartbeatFailureDetector.java

```diff
@@ -143,6 +147,31 @@ public void shutdown()
                 .collect(toImmutableSet());
     }
 
+    @Override
+    public State getState(HostAddress hostAddress)
+    {
+        for (MonitoringTask task : tasks.values()) {
+            if (hostAddress.equals(HostAddress.fromUri(task.uri))) {
+                if (!task.isFailed()) {
+                    return State.ALIVE;
+                }
+                else {
+                    Exception lastException = task.getStats().getLastFailureException();
+                    if (lastException instanceof SocketTimeoutException || lastException instanceof UnknownHostException) {
+                        return State.GONE;
+                    }
+                    else if (lastException instanceof ConnectException) {
+                        return State.UNRESPONSIVE;
+                    }
+                    else {
+                        return State.UNKNOWN;
```

Comment:
Thank you for these thoughts!! :)

- Today the `tasks` is a map from `UUID` (get from the `ServiceDescriptor`) to the `MonitoringTask`. I didn't see an easy way to reverse lookup the `ServiceDescriptor` based on hostName. 

- It's not common but it does happen. Here are something I am currently aware of:
  - Something like `org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation` will be returned if we ping on a non-HTTP port (e.g. port 22)
  - Some kind of "EOFException" will be returned when the machine is restarting, or some weird firewall issue that allows the connection but doesn't allow any communication.